WIP feat: add garbage collection to notion workflow

fontanierh opened the pull request.

fontanierh commented on connectors/src/connectors/notion/temporal/activities.ts:
```diff
@@ -152,3 +165,208 @@ export async function getNotionAccessTokenActivity(
   return notionAccessToken;
 }
+
+export async function shouldGarbageCollectActivity(
```
cc @spolu  this is the function where we decide if a given DS should go into "garbage collection" mode.
Maybe I should add something to make sure it happens during the night (but we don't have user timezones yet :/) ? For a big workspace, the process halts synchronization for ~25mn

spolu replied:
That seems fine for now :+1:

fontanierh commented on connectors/src/connectors/notion/temporal/workflows.ts:
```diff
@@ -89,26 +100,45 @@ export async function notionSyncWorkflow(
       pageIndex += 1;
 
-      const pagesToSync = pageIds.filter(
-        (pageId) => !pagesSyncedWithinPeriod.has(pageId)
```
got rid of this "optimization": we sync every ~10s, but notion TS is only precise to the minute, which means when a page changes we might end up upserting it up to 6 times. I think it is better than potentially missing changes (eg, the page changed multiple times during a minute, and stops updating before the end of the minute).

If we end up upserting too much, we can consider increasing the interval between our sync runs

fontanierh commented on connectors/src/connectors/notion/temporal/workflows.ts:
```diff
@@ -133,21 +165,37 @@ export async function notionSyncWorkflow(
+      lastSyncedPeriodTs = runTimestamp;
+    } else {
+      // garbage collect the pages -- can be done non-blocking
+      void executeChild(notionGarbageCollectWorkflow.name, {
```
the only reason this is a workflow and not an activity is because I don't want to wait for it to complete in the parent WF. The parent can be cont'd as new regardless of the status of the child

spolu commented on connectors/src/lib/models.ts:
```diff
@@ -96,6 +100,18 @@ Connector.init(
       type: DataTypes.DATE,
       allowNull: true,
     },
+    lastGarbageCollectionStartTime: {
```
@lasryaric r? for addition to the connector's model

spolu replied:
@lasryaric bump r?

fontanierh commented on connectors/src/lib/models.ts:
```diff
@@ -96,6 +100,18 @@ Connector.init(
+      type: DataTypes.DATE,
+      allowNull: true,
+    },
+    firstSuccessfulSyncTime: {
```
yes, it will be added automatically (it will not be the true first sync but it doesn't matter)

fontanierh commented:
@spolu 

> a bit 🔥 to share the workflow but OK

I hesitated, but I don't want the other workflow to keep syncing while it's garbage collecting (rate limit etc..) so I figured it was easier like this. _technically_ we could continue syncing while this happens though ?

spolu commented:
Agreed that the approach is reasonable :+1:

fontanierh commented:
@spolu  @lasryaric PTAL

- I addressed @spolu  comments
- I added some code to avoid re-processing pages that are already up-to-date (i.e our "last visited" TS in the DB is >= to notion's `last_edited` TS)

spolu reviewed (dismissed):
LGTM, but would like to wait on @lasryaric r? on the models changes before merging :+1:

fontanierh commented:
Deploy:
- `npm run initdb`
- `env $(cat .env.local) npx tsx ./migrations/20230505_add_notion_connector_state.ts`
- deploy worker
- `npm run cli -- notion restart-all`

spolu commented on connectors/src/lib/models.ts:
```diff
@@ -149,6 +156,58 @@ SlackConfiguration.init(
+
+Connector.hasOne(NotionConnectorState, {
+  as: "notionConnectorState",
+  foreignKey: "connectorId",
```
not needed right?

fontanierh replied:
I don't know why but it seemed to be needed

spolu replied:
Well not needed for the SlackConfiguration. Maybe this is because you define option object?

fontanierh replied:
SlackConfiguration doesn't use the automatic joins feature of the ORM

spolu replied:
My point exactly ;-)

spolu commented on connectors/src/connectors/notion/temporal/activities.ts:
```diff
@@ -152,3 +249,201 @@ export async function getNotionAccessTokenActivity(
+      throw new Error("Could not find notionConnectorState");
+    }
+
+    await connector.notionConnectorState.update({
```
I'm not a suppper fan of the result here, feels too magical to me. But this is a weakly held position, so happy to run with it.

spolu replied:
(Still weakly held)

fontanierh replied:
I'm fine with that, I don't mind the second query

spolu commented on connectors/src/connectors/notion/temporal/activities.ts:
```diff
@@ -94,10 +175,13 @@ export async function saveSuccessSyncActivity(
 
     const now = new Date();
 
+    const firstSuccessfulSyncTime = connector.firstSuccessfulSyncTime || now;
```
This seems like something we need to handle on the slack size as well no?

fontanierh replied:
Added to the `syncStatus` function that are shared with slack (wasn't aware of those previously)
https://github.com/dust-tt/dust/pull/390/commits/77d5e2132601ef2995934f09210b180a2cc33d6e
cc @lasryaric

spolu reviewed (approved):
One big question on the impact of the new field on Connector for slack since it seems that you have Notion specific code to handle it that is was not ported to slack?

Also r? @lasryaric for the models change.

Otherwise assuming this is addressed, LGTM
